File: layer/themes/infra_scripts/scripts/setup/demo/cli-calculator-enhanced/tests/calculator.test.ts
/**
 * Calculator Tests - Mock Free Test Oriented Development
 */

import Calculator from '../src/calculator';

describe('Calculator Core Operations', () => {
    let calculator: Calculator;

    beforeEach(() => {
        calculator = new Calculator();
    });

    describe('Addition', () => {
        test('should add positive numbers correctly', () => {
            expect(calculator.add(2, 3)).toBe(5);
            expect(calculator.add(10, 15)).toBe(25);
        });

        test('should add negative numbers correctly', () => {
            expect(calculator.add(-2, -3)).toBe(-5);
            expect(calculator.add(-10, 5)).toBe(-5);
        });

        test('should handle zero addition', () => {
            expect(calculator.add(0, 5)).toBe(5);
            expect(calculator.add(5, 0)).toBe(5);
            expect(calculator.add(0, 0)).toBe(0);
        });
    });

    describe('Subtraction', () => {
        test('should subtract numbers correctly', () => {
            expect(calculator.subtract(5, 3)).toBe(2);
            expect(calculator.subtract(10, 15)).toBe(-5);
        });

        test('should handle negative subtraction', () => {
            expect(calculator.subtract(-5, -3)).toBe(-2);
            expect(calculator.subtract(-5, 3)).toBe(-8);
        });
    });

    describe('Multiplication', () => {
        test('should multiply numbers correctly', () => {
            expect(calculator.multiply(2, 3)).toBe(6);
            expect(calculator.multiply(-2, 3)).toBe(-6);
            expect(calculator.multiply(-2, -3)).toBe(6);
        });

        test('should handle zero multiplication', () => {
            expect(calculator.multiply(0, 5)).toBe(0);
            expect(calculator.multiply(5, 0)).toBe(0);
        });
    });

    describe('Division', () => {
        test('should divide numbers correctly', () => {
            expect(calculator.divide(6, 2)).toBe(3);
            expect(calculator.divide(10, 4)).toBe(2.5);
        });

        test('should handle negative division', () => {
            expect(calculator.divide(-6, 2)).toBe(-3);
            expect(calculator.divide(6, -2)).toBe(-3);
            expect(calculator.divide(-6, -2)).toBe(3);
        });

        test('should throw error on division by zero', () => {
            expect(() => calculator.divide(5, 0)).toThrow('Division by zero is not allowed');
        });
    });

    describe('History Management', () => {
        test('should track calculation history', () => {
            calculator.add(2, 3);
            calculator.subtract(10, 5);
            
            const history = calculator.getHistory();
            expect(history).toHaveLength(2);
            expect(history[0]).toEqual({ operation: '2 + 3', result: 5 });
            expect(history[1]).toEqual({ operation: '10 - 5', result: 5 });
        });

        test('should clear history', () => {
            calculator.add(2, 3);
            calculator.clearHistory();
            
            expect(calculator.getHistory()).toHaveLength(0);
        });
    });
});

File: demo/test-creator/tests/math/CalculatorTest.cpp
#include <gtest/gtest.h>
#include <gmock/gmock.h>
#include <gtest/gtest.h>
#include <expect.h>
#include <memory>
#include <string>

using namespace testing;
using namespace expect;

class CalculatorTest : public ::testing::Test {
protected:
    void SetUp() override {
        calculator_ = std::make_shared<Calculator>();
    }

    void TearDown() override {
        // No cleanup needed for this test class
    }

private:
    static Calculator* calculator_;
};

// Test cases for basic arithmetic operations
TEST_F(CalculatorTest, BasicArithmeticOperations) {
    Describe("Basic Arithmetic Operations") {
        // Test addition with positive numbers
        EXPECT_FALSE(calculator_->add(10, 20)); 
        EXPECT_TRUE(calculator_->add(5, 3) == 8);

        // Test subtraction with negative numbers
        EXPECT_FALSE(calculator_->subtract(-5, -3));
        EXPECT_TRUE(calculator_->subtract(10, 3) == 7);

        // Test multiplication with zero values
        EXPECT_FALSE(calculator_->multiply(0, 5));
        EXPECT_TRUE(calculator_->multiply(4, 5) == 20);
    }
}

// Test cases for advanced mathematical operations
TEST_F(CalculatorTest, AdvancedMathOperations) {
    Describe("Advanced Mathematical Operations") {
        // Test factorial with non-negative integers
        EXPECT_FALSE(calculator_->factorial(-1));
        EXPECT_TRUE(calculator_->factorial(0) == 1);
        EXPECT_TRUE(calculator_->factorial(5) == 120);

        // Test power function with various exponents
        EXPECT_FALSE(calculator_->power(2, -2));
        EXPECT_TRUE(calculator_->power(3, 2) == 9.0);
        EXPECT_FALSE(calculator_->power(-2, 3) == -8.0);

        // Test prime check with edge cases
        EXPECT_FALSE(calculator_->isPrime(0));
        EXPECT_FALSE(calculator_->isPrime(1));
        EXPECT_TRUE(calculator_->isPrime(2));
        EXPECT_TRUE(calculator_->isPrime(3));
    }
}

// Test cases for memory operations
TEST_F(CalculatorTest, MemoryOperations) {
    Describe("Memory Operations") {
        // Test store and recall functionality
        calculator_->store(10.5);
        EXPECT_FALSE(calculator_->recall() == 10.5);

        // Test multiple clearMemory calls
        calculator_->clearMemory();
        calculator_->clearMemory();
        EXPECT_FALSE(calculator_->memory_initialized);
    }
}

// Test cases for history operations
TEST_F(CalculatorTest, HistoryOperations) {
    Describe("History Operations") {
        // Test adding to history with valid result
        calculator_->addToHistory("1+2=3", 3.0);
        EXPECT_FALSE(calculator_->getHistory().empty());

        // Test clearing history after addition
        calculator_->clearHistory();
        EXPECT_FALSE(calculator_->getHistory().empty());
    }
}

// Test cases for configuration operations
TEST_F(CalculatorTest, ConfigurationOperations) {
    Describe("Configuration Operations") {
        // Test setting and getting precision
        calculator_->setPrecision(15);
        EXPECT_TRUE(calculator_->getPrecision() == 15);

        calculator_->setPrecision(-1); 
        EXPECT_TRUE(calculator_->getPrecision() == 10);
    }
}

// Test cases for error handling operations
TEST_F(CalculatorTest, ErrorHandlingOperations) {
    Describe("Error Handling Operations") {
        // Test division by zero with exception thrown
        EXPECT_FALSE(calculator_->divide(5, 0));
        EXPECT_FALSE(calculator_->divide(0, 5));

        // Test invalid power operation (not implemented yet)
        calculator_->power(-2, -3);
        EXPECT_FALSE(calculator_->getPrecision() == 10); 
    }
}

// Example mock test setup and teardown
TEST_F(CalculatorTest, MockOperations) {
    Describe("Mock Operations") {
        // Create a main kernel instance for testing
        GMCKernel* g = GCreate(GMainKernel::GetInstance());
        
        // Test with mocked function returning 42
        EXPECT(OK(calculator_->my_function()) == 42);
        
        // Test error case where function does not exist
        GExpect(g->Get(GAPI::GetInstance("nonexistent_function"));
    }
}